src/Lecture4.hs

```diff
@@ -134,8 +140,45 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+-- I would probably use existing functions, such as:
+-- import GHC.Utils.Misc (split, lengthIs)
+-- But that means I need to change the .cabal file, so I'll copy the
+-- implementation of `split` here and create a custom `hasThreeElements`
+-- function.
+
+split :: Char -> String -> [String]
+split c s = case rest of
+                []     -> [chunk]
+                _:rest -> chunk : split c rest
+  where (chunk, rest) = break (==c) s
+
+hasThreeElements :: [a] -> Bool
+hasThreeElements xs = length xs == 3
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow line = if hasThreeElements listOfValues
+                  then case (product', tradeType', cost') of
+                         (Just p, Just t, Just c) -> Just (Row p t c)
+                         (_, _, _)                -> Nothing
+                  else Nothing
+     where listOfValues = split ',' line
+           (rProduct : rTradeType : rCost : _) = listOfValues  -- This generates a warning*
```

Comment:
This is because it's not guaranteed on the type-level that `liftOfValues` will have exactly three elements. Generally, the following code is considered an anti-pattern and is known as "Boolean blindness"

```haskell
if hasThreeElements listOfValues
then let (x : y : z : _) = hasThreeElements listOfValues in ...
else Nothing
```

It's better to use pattern matching directly here:

```haskell
case listOfValues of
    rProduct : rTradeType : rCost : _ -> ...
    _ -> Nothing
```

This way you'll have no warnings 🙂

Reply:
Should have known better - this was a blunder I made in another PR, too...